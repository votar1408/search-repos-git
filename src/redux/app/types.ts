export const LOADING: string = 'LOADING';

export interface AppLoadingAction {
    type: typeof LOADING;
    payload: boolean;
}

export type AppAction = AppLoadingAction;

export interface AppState {
    loading: boolean;
}
