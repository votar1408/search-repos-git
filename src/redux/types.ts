import {AppState} from './app/types';

export interface RootAppState {
    app: AppState;
}

interface RootState extends RootAppState {}

export const selectApp = (state: RootState) => state.app;
