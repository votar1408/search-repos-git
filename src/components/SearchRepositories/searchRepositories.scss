@import '../../assets/scss/variable.scss';

.search {
    min-width: 40%;
    position: relative;
    display: flex;
    margin: 18px 0;
}

.search__term {
    width: 100%;
    border: 2px solid $color-main;
    border-right: none;
    padding: 5px;
    border-radius: 5px 0 0 5px;
    outline: none;
    font-size: 16px;
}

.search__button {
    width: 40px;
    height: 36px;
    border: 2px solid $color-main;
    background: $color-main;
    text-align: center;
    color: #fff;
    border-radius: 0 5px 5px 0;
    cursor: pointer;
    font-size: 20px;
    transition: 1s;
    &:hover {
        background: lighten($color: $color-main, $amount: 10);
    }
}
