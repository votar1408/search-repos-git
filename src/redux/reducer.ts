import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';

export const reducer = combineReducers({
    app: appReducer
});
