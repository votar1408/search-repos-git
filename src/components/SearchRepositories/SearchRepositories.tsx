import './searchRepositories.scss';

import React, {
    ChangeEvent,
    FC,
    FormEvent,
    useCallback,
    useEffect,
    useState
} from 'react';

import {
    useDispatch,
    useSelector
} from 'react-redux';

import {
    common,
    texts
} from '../../consts';
import {
    changeSearchValueAction,
    clearRepositoriesAction,
    getRepositoriesAction
} from '../../redux/app/actions';
import {selectApp} from '../../redux/types';
import {getRepositoriesUrl} from '../../routes';
import {
    useHttp,
    useMountedRef
} from '../hooks';

export const SearchRepositories: FC = () => {
    const dispatch = useDispatch();
    const isMountedRef = useMountedRef();
    const {page, repositories, searchValue} = useSelector(selectApp);
    const {request, controller} = useHttp();
    const [searchValueState, setSearchValueState] = useState<string>('');

    useEffect(() => {
        onSendRequestHandler();
        return () => controller.abort();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [page]);

    const onSendRequestHandler = useCallback(async () => {
        const isNewSearchValue: boolean = searchValueState !== searchValue;
        if (searchValueState && (isNewSearchValue || !repositories[page])) {
            try {
                const data = await request(
                    getRepositoriesUrl(searchValueState, isNewSearchValue ? common.getDefaultPage : page),
                    'GET'
                );

                if (isMountedRef.current) {
                    dispatch(getRepositoriesAction(data.items));
                }
            } catch (e) {
                alert(e.message);
            }
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [dispatch, searchValue, searchValueState, page, repositories]);

    const onChangeSearchValueHandler = useCallback((event: ChangeEvent<HTMLInputElement>) => {
        event.persist();
        setSearchValueState(event.target.value);
    }, []);

    const onSearchRepositoriesHandler = useCallback(
        async (event: FormEvent<HTMLFormElement>) => {
            event.preventDefault();

            onSendRequestHandler();
            if (isMountedRef.current) {
                if (searchValueState !== searchValue) {
                    dispatch(changeSearchValueAction(searchValueState));
                    dispatch(clearRepositoriesAction());
                }
            }
        },
        [dispatch, isMountedRef, onSendRequestHandler, searchValue, searchValueState]
    );

    return (
        <form className="search" onSubmit={onSearchRepositoriesHandler}>
            <input
                type="text"
                name="search-repo-input"
                className="search__term"
                placeholder={texts.inputRepositoryName}
                value={searchValueState}
                onChange={onChangeSearchValueHandler}
            />
            <button type="submit" name="submit-button" className="search__button">
                <i className="fa fa-search"></i>
            </button>
        </form>
    );
};
