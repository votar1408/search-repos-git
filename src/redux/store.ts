/* eslint-disable no-underscore-dangle */
import {
    applyMiddleware,
    compose,
    createStore,
    Middleware
} from 'redux';
import thunk from 'redux-thunk';

import {reducer} from './reducer';

const middleware: [Middleware] = [thunk];

export const store = createStore(reducer, compose(applyMiddleware(...middleware)));
