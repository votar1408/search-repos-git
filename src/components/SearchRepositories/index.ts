import {SearchRepositories} from './SearchRepositories';

export {SearchRepositories};
