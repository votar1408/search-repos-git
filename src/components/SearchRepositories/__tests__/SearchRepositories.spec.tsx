import React from 'react';

import {
    mount,
    ReactWrapper,
    ShallowWrapper
} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import createMockStore, {
    MockStoreCreator,
    MockStoreEnhanced
} from 'redux-mock-store';

import {RootAppState} from '../../../redux/types';
import {SearchRepositories} from '../SearchRepositories';

describe('SearchRepositories', () => {
    const initialState: RootAppState = {
        app: {
            loading: false,
            repositories: [],
            page: 1,
            searchValue: ''
        }
    };

    const mockStore: MockStoreCreator = createMockStore();

    let store: MockStoreEnhanced;
    let wrapper:
        | ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>
        | ShallowWrapper<{}, {}, React.Component<{}, {}, any>>;

    const getWrapper = (state: RootAppState) => {
        store = mockStore(state);
        return mount(
            <Router>
                <Provider store={store}>
                    <SearchRepositories />
                </Provider>
            </Router>
        );
    };

    beforeEach(() => {
        wrapper = getWrapper(initialState);
    });

    it('should render correctly', () => {
        expect(shallowToJson(wrapper as ShallowWrapper)).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        expect(wrapper.find(SearchRepositories).length).toEqual(1);
    });

    it('should have 1 input element', () => {
        expect(wrapper.find('input').length).toEqual(1);
    });

    it('should update value for field after some input', () => {
        let searchInput = wrapper.find('input[name="search-repo-input"]');
        expect(searchInput.prop('value')).toEqual('');

        const expectValue = {
            target: {
                name: 'search-repo-input',
                value: 'new some value'
            }
        };
        searchInput.simulate('change', expectValue);

        searchInput = wrapper.find('input[name="search-repo-input"]');
        expect(searchInput.prop('value')).toEqual(expectValue.target.value);
    });

    it('should update value for field after some input', () => {
        let searchInput = wrapper.find('input[name="search-repo-input"]');
        expect(searchInput.prop('value')).toEqual('');

        const expectValue = {
            target: {
                name: 'search-repo-input',
                value: 'new some value'
            }
        };
        searchInput.simulate('change', expectValue);

        searchInput = wrapper.find('input[name="search-repo-input"]');
        expect(searchInput.prop('value')).toEqual(expectValue.target.value);
    });

    it('should enabled submit button', () => {
        const submitButton = wrapper.find('button[name="submit-button"]');

        submitButton.simulate('click');

        expect(submitButton.prop('disabled')).toBeFalsy();
    });
});
