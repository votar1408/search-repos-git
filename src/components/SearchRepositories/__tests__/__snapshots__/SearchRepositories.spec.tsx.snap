// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchRepositories should render correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <SearchRepositories>
        <form
          className="search"
          onSubmit={[Function]}
        >
          <input
            className="search__term"
            name="search-repo-input"
            onChange={[Function]}
            placeholder="Input the name of the repository"
            type="text"
            value=""
          />
          <button
            className="search__button"
            name="submit-button"
            type="submit"
          >
            <i
              className="fa fa-search"
            />
          </button>
        </form>
      </SearchRepositories>
    </Provider>
  </Router>
</BrowserRouter>
`;
